packages/auth/src/auth.interfaces.ts

```diff
@@ -31,14 +48,34 @@ export interface ShopifyAuthAfterHandler<
   afterAuth(req: T, res: R, session: Session): Promise<void>;
 }
 
-export interface ShopifyAuthModuleAsyncOptions
-  extends Pick<ModuleMetadata, 'imports'> {
-  useExisting?: Type<ShopifyAuthOptionsFactory>;
-  useClass?: Type<ShopifyAuthOptionsFactory>;
+export interface ShopifyAuthTokenExchangeAfterHandlerParams {
+  session: Session;
+  sessionToken: string;
+}
+export interface ShopifyAuthTokenExchangeAfterHandler {
+  afterAuth(params: ShopifyAuthTokenExchangeAfterHandlerParams): Promise<void>;
+}
+
+export type AuthStrategy = 'TOKEN_EXCHANGE' | 'AUTHORIZATION_CODE_FLOW';
+export type ShopifyAuthModuleAsyncOptions<
+  A extends AuthStrategy,
+  O = A extends 'AUTHORIZATION_CODE_FLOW'
+    ? ShopifyAuthModuleAuthorizationCodeFlowBaseOptions
+    : ShopifyAuthModuleTokenExchangeBaseOptions,
+> = Pick<ModuleMetadata, 'imports'> & {
+  authStrategy?: A;
+  useExisting?: Type<ShopifyAuthOptionsFactory<O>>;
+  useClass?: Type<ShopifyAuthOptionsFactory<O>>;
   /* eslint-disable @typescript-eslint/no-explicit-any */
-  useFactory?: (
-    ...args: any[]
-  ) => Promise<ShopifyAuthModuleOptions> | ShopifyAuthModuleOptions;
+  useFactory?: (...args: any[]) => Promise<O> | O;
   inject?: any[];
   /* eslint-enable @typescript-eslint/no-explicit-any */
+};
+
+export interface ShopifyAuthStrategyService {
```

Comment:
Can you make this an `abstract class`?

The neat thing of abstract classes is that you can still use `class A implements AbstractClassNameHere`. But the other nice part is that you can use it in the provider definition:

```typescript
{
  provide: ShopifyAuthStrategyService,
  useClass: ShopifyTokenExchangeAuthStrategyOfflineService,
}
```

or

```typescript
{
  provide: ShopifyAuthStrategyService,
  useClass: ShopifyTokenExchangeAuthStrategyOfflineService,
}
```

etc.

Reply:
I just implemented this.